### PR TITLE
Pr.pkg.corrade

### PR DESCRIPTION
**Pull request by pthom**
* I have checked that this pull request contains only
  `.travis.yml`/`appveyor.yml` changes. All other changes send
  to https://github.com/ruslo/hunter. **[Yes]**

Note : this PR also contains a new script 
magnum-build-corrade-rc.sh : I explain the reason here : https://github.com/ruslo/hunter/issues/1635

Summary :  
1. concerning the cross platform builds, corrade needs a native exe (corrade-rc) to be built before building the library. So I had to modify .travis.yml in order to first build it.
2. I also switched travis base distrib from ubuntu 14 (trusty) to 16 (xenial). I had to do this because there were some compiler errors, and also because OpenGL support is always better with more recent distributions

* I have checked that no toolchains removed from CI configs, they are commented
  out instead so other developers can enable them back easily and to simplify
  merge conflict resolution. **[Yes]**

* I have checked that for every commented out toolchain there is a link to the
  broken CI build page or to the minimum compiler requirements documentation
  so other developers can figure out what was the problem exactly. **[Yes]**

* Links to succesful builds: 
https://ci.appveyor.com/project/pthom/hunter/builds/20737216
https://travis-ci.org/pthom/hunter/builds/462950117

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -134,6 +145,11 @@ install:
   - export ANDROID_NDK_r17="`pwd`/_ci/android-ndk-r17"
 
 script:
+  - | # build corrade-rc if cross compiling
+    if [ ! -z ${CROSS_COMPILE+x} ]; then
+      ./magnum-build-corrade-rc.sh
```
Comment (ruslo) [inline .travis.yml]: Please try to use the `hunter_experimental_add_host_project` feature first, see https://github.com/ruslo/hunter/issues/1635#issuecomment-443826466

File: .travis.yml [hunk 2]
```diff
@@ -23,7 +23,7 @@ addons:
       - g++-7
 
 dist:
-  - trusty
+  - xenial
```
Comment (ruslo) [inline .travis.yml:26]: Will test it, looks like `trusty` -> `xenial` "just work" 👍

**Review by ruslo — CHANGES_REQUESTED** [review]

**Comment by mosra** [comment]
> I also switched travis base distrib from ubuntu 14 (trusty) to 16 (xenial). I had to do this because there were some compiler errors, and also because OpenGL support is always better with more recent distributions

I think this could still work with 14.04. Was there anything else except the error fixed in https://github.com/mosra/corrade/commit/ae1fc17d3d77b1bec248382fbdbefd8b9fbb44a9? Besides that, Magnum is quite independent on system OpenGL support -- it uses its own headers and fetches function pointers dynamically, so driver version at compile time doesn't matter at all, only driver version at runtime.

File: .travis.yml (hunk 2 above)
Comment (ruslo) [reply .travis.yml:26]: Applied globally: https://github.com/ingenue/hunter/commit/1fbc0423cacd3a2d6a60b9052183982c8847dc70

**Comment by pthom** [comment]
@mosra  : 
> I think this could still work with 14.04

Yes it can probably. I pushed a commit that goes back to travis. However, I hope not to have to revert this later ;-)
See https://travis-ci.org/pthom/hunter/builds/463004762 (android does not build, it is another issue, discussed in this thread)

File: .travis.yml (hunk 1 above)
Comment (pthom) [reply .travis.yml]: Well, I cannot say I did not try 

See my commits here : https://github.com/pthom/hunter/commit/ff01c8840f40e29818d54b683bcfec33376ba74a

See the build log :
- Step 1 :  https://travis-ci.org/pthom/hunter/jobs/463004765#L1184 
Here this is `hunter/examples/corrade/CMakeLists.txt` speaking
`````cmake
if(IOS OR ANDROID)
  # add cmake/host subdiretcory as host project to install protoc
  include(hunter_experimental_add_host_project)
  hunter_experimental_add_host_project(cmake/host)

  add_executable(Corrade::corrade-rc IMPORTED)
  set_property(TARGET Corrade::corrade-rc APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
  set_target_properties(Corrade::corrade-rc PROPERTIES IMPORTED_LOCATION_RELEASE "${HUNTER_HOST_ROOT}/bin/corrade-rc")

  set(CORRADE_RC_EXECUTABLE "${HUNTER_HOST_ROOT}/bin/corrade-rc")

  if(EXISTS ${CORRADE_RC_EXECUTABLE})
    message("Found CORRADE_RC_EXECUTABLE at ${CORRADE_RC_EXECUTABLE}")
  else()
    message("Did not find CORRADE_RC_EXECUTABLE at ${CORRADE_RC_EXECUTABLE}")
  endif()
  message(STATUS "Using imported corrade-rc from host: ${HUNTER_HOST_ROOT}/bin/corrade-rc")
endif(IOS OR ANDROID)

`````

- Step 2 :  https://travis-ci.org/pthom/hunter/jobs/463004765#L1383

Here, it is `corrade/CMakeLists.txt` that speaks:

````cmake
if(CMAKE_CROSSCOMPILING)
    find_program(CORRADE_RC_EXECUTABLE corrade-rc)
    if(NOT CORRADE_RC_EXECUTABLE)
        message(FATAL_ERROR "Native `corrade-rc` executable, which is needed when crosscompiling, was not found")
    endif()
    add_executable(Corrade::rc IMPORTED)
    set_property(TARGET Corrade::rc PROPERTY IMPORTED_LOCATION ${CORRADE_RC_EXECUTABLE})
endif()
````

I'm out of ideas. Are you?

**Comment by pthom** [comment]
@mosra  @ruslo  : I downgraded to trusty, then saw that a global move to xenial by @ruslo 
-> I will upgrade to xenial again.
See : https://travis-ci.org/pthom/hunter/builds/463029503

File: .travis.yml (hunk 2 above)
Comment (pthom) [reply .travis.yml:26]: Warning: it does not "just work". 
For example, I had to add `libgl1-mesa-dev` which was not by default in xenial.
See
https://github.com/pthom/hunter/commit/ca96d92dc2ff5cdf2e3f8632d17297b222589689

**Comment by mosra** [comment]
@pthom oh. sorry about the extra work, then

File: .travis.yml (hunk 1 above)
Comment (ruslo) [reply .travis.yml]: > I'm out of ideas. Are you?

> https://github.com/pthom/hunter/commit/ff01c8840f40e29818d54b683bcfec33376ba74a

This code should be part of "corrade" sources, not in examples. User don't have to do anything special, everything should "just work" with `hunter_add_package(corrade)` even if you're cross-compiling.

See my comment about ICU package: https://github.com/ruslo/hunter/issues/1635#issuecomment-443826466

On user side everything still plain simple even for iOS/Android:
* https://github.com/ruslo/hunter/blob/v0.23.63/examples/ICU/CMakeLists.txt#L14-L17

File: .travis.yml (hunk 2 above)
Comment (ruslo) [reply .travis.yml:26]: > For example, I had to add libgl1-mesa-dev which was not by default in xenial

That's package specific I guess.

File: .travis.yml (hunk 1 above)
Comment (ruslo) [reply .travis.yml]: You can skip iOS/Android testing for now and deal with it later.

**Comment by pthom** [comment]
Ok, I removed the iOS and Android builds.
Successful builds here: 
https://ci.appveyor.com/project/pthom/hunter/builds/20781352
and 
https://travis-ci.org/pthom/hunter/builds/463731734

**Review by ruslo — APPROVED** [review]

**Pull request merged**